File: switsol_ag/custom_scripts/address.js
// Copyright (c) 2017, Switsol AG and contributors
// For license information, please see license.txt
cur_frm.add_fetch("postal_code", "canton", "canton");
cur_frm.add_fetch("town", "canton", "canton")

frappe.ui.form.on("Address", {
	onload: function(frm) {
		frm.set_query("country", function() {
			return {
				filters: {
					"flag": 1
				}
			};
		});
		frm.set_query("postal_code", function(doc, cdt, cdn) {
			if(frm.doc.town) {
				return {
					query: "switsol_ag.address.postal_code_query",
					filters: {
						"town": frm.doc.town
					}
				}
			}
		});
		frm.set_query("town", function(doc, cdt, cdn) {
			if(frm.doc.postal_code) {
				return {
					query: "switsol_ag.address.city_query",
					filters: {
						"postal_code": frm.doc.postal_code
					}
				}
			}
		});
	},
	refresh: function(frm) {
		if(frappe.route_options) {
			cur_frm.set_value("address_title", frappe.route_options.address_title);
			cur_frm.set_value("customer", frappe.route_options.customer);
			cur_frm.set_value("address_type", "Billing");
			cur_frm.set_value("country", frappe.boot.sysdefaults.country);
		}
	},
	town: function(frm) {
		frappe.call({
			method:"frappe.client.get_list",
			args: {
				doctype:"City Postal Code",
				filters: [
					["parent","=", frm.doc.town]
				],
				fields: ["postal_code"]
			},
			callback: function(r) {
				if(r.message) {
					postal_codes = [];
					$.each(r.message, function(j, data) {
						postal_codes.push(data.postal_code);
					})
					if(postal_codes.length == 1) {
					 	frm.set_value("postal_code", postal_codes[0]);
					}
				}
			}
		})
	}
});

frappe.ui.form.on("Address", "add_customers", function(frm) {
	var dialog = new frappe.ui.Dialog({
		title: __("Add Reference"),
		fields: [
			{fieldtype: "Link", fieldname: "customer", label: __("Customer"), reqd: 1, options: "Customer"}
		]
	});

	dialog.set_primary_action(__("Add"), function() {
		var btn = this;
		var values = dialog.get_values();
		frappe.call({
			freeze: true,
			freeze_message: "Saving document in the background",
			type: "POST",
			method: "switsol_ag.address.add_customers",
			args: {
				address: frm.doc.name,
				customer: values.customer,
			},
			callback: function(r) {
				if(!r.exc) {
					cur_frm.reload_doc();
				}
			}
		});
		dialog.hide();
	});
	dialog.show();
});

frappe.ui.form.on("Address", "remove_customers", function(frm) {
	var dialog = new frappe.ui.Dialog({
		title: __("Delete Reference"),
		fields: [
			{fieldtype: "Link", fieldname: "customer", label: __("Customer"), reqd: 1, options: "Customer"}
		]
	});

	dialog.set_primary_action(__("Delete"), function() {
		var btn = this;
		var values = dialog.get_values();
		frappe.call({
			freeze: true,
			freeze_message: "Saving document in the background",
			type: "POST",
			method: "switsol_ag.address.remove_customers",
			args: {
				address: frm.doc.name,
				customer: values.customer,
			},
			callback: function(r) {
				if(!r.exc) {
					cur_frm.reload_doc();
				}
			}
		});
		dialog.hide();
	});
	dialog.show();
});

File: switsol_ag/address.py
# Copyright (c) 2017, Switsol AG and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.utils import get_link_to_form


def validate(self, method=None):
    self.pincode = self.postal_code.strip() if self.postal_code else ""
    if self.customer and self.is_primary_address:
        frappe.db.set_value("Customer", self.customer, "pincode", self.postal_code)


@frappe.whitelist()
def postal_code_query(doctype, txt, searchfield, start, page_len, filters):
    return frappe.db.sql("""select postal_code
        from `tabCity Postal Code`
        where parent=%s
        and postal_code like %s
        order by postal_code desc
        limit %s, %s""".format(frappe.db.escape(searchfield)), (filters.get("town"), "%{0}%".format(txt), start, page_len))


@frappe.whitelist()
def city_query(doctype, txt, searchfield, start, page_len, filters):
    return frappe.db.sql("""select cp.parent
        from `tabCity Postal Code` cp
        where cp.postal_code=%s
        order by parent desc
        limit %s, %s""".format(frappe.db.escape(searchfield)), (filters.get("postal_code"), start, page_len))


@frappe.whitelist()
def add_customers(customer, address):
    customer_name = frappe.db.sql("""select customer_name
        from tabCustomer
        where name=%s""", customer, as_dict=1)
    address = frappe.db.sql("""select name, customers
        from tabAddress
        where name=%s""", address, as_dict=1)
    if customer_name and address:
        customers = customer_name[0].customer_name + " " + get_link_to_form("Customer", customer, customer)
        if address[0].customers:
            customers_name = [i.split('<a')[0].strip() for i in address[0].customers.split('\n')]
            if customer_name[0].customer_name not in customers_name:
                customers = address[0].customers + "\n" + customer_name[0].customer_name + " " + get_link_to_form("Customer", customer, customer)
            else:
                customers = address[0].customers
        frappe.db.set_value("Address", address[0].name, "customers", customers)
        frappe.db.commit()


@frappe.whitelist()
def remove_customers(customer, address):
    customer_name = frappe.db.sql("""select customer_name
        from tabCustomer
        where name=%s""", customer, as_dict=1)
    address = frappe.db.sql("""select name, customers
        from tabAddress
        where name=%s""", address, as_dict=1)
    if customer_name and address:
        if address[0].customers:
            customers = address[0].customers.split('\n')
            for i in customers:
                if i.find(customer_name[0].customer_name) != -1:
                    customers.remove(i)
            frappe.db.set_value("Address", address[0].name, "customers", '\n'.join(customers))
            frappe.db.commit()


@frappe.whitelist()
def make_address(company_name):
    ad = frappe.new_doc("Address")
    ad.customer = company_name
    ad.address_title = company_name
    return ad
